interpreter/exec/eval.ml

```diff
@@ -441,23 +456,70 @@ let rec step_thread (t : thread) : thread =
         let locals' = List.rev args @ List.map default_value f.it.locals in
         let frame' = {inst = !inst'; locals = List.map ref locals'} in
         let instr' = [Label (n2, [], ([], List.map plain f.it.body)) @@ f.at] in
-        vs', [Frame (n2, frame', ([], instr')) @@ e.at]
+        vs', [Frame (n2, frame', ([], instr')) @@ e.at], NoAction
 
       | Func.HostFunc (t, f) ->
-        try List.rev (f (List.rev args)) @ vs', []
+        try List.rev (f (List.rev args)) @ vs', [], NoAction
         with Crash (_, msg) -> Crash.error e.at msg
       )
-  in {t with code = vs', es' @ List.tl es}
 
+    | Suspend _, vs ->
+      (* TODO: meaningful timestamp handling *)
+      vs, [e], NoAction
+
+  in {t with code = vs', es' @ List.tl es}, act
+
+let rec plug_value (c : code) (v : value) : code =
+  let vs, es = c in
+  match es with
+  | {it = Label (n, es0, c'); at} :: es' ->
+    vs, {it = Label (n, es0, plug_value c' v); at} :: es'
+  | {it = Frame (n, f, c'); at} :: es' ->
+    vs, {it = Frame (n, f, plug_value c' v); at} :: es'
+  | _ ->
+    v :: vs, es
+
+let rec try_unsuspend (c : code) (m : memory_inst) (addr : Memory.address) : code option =
+  let vs, es = c in
+  match es with
+  | {it = Label (n, es0, c'); at} :: es' ->
+    Lib.Option.map (fun c'' -> vs, {it = Label (n, es0, c''); at} :: es') (try_unsuspend c' m addr)
+  | {it = Frame (n, f, c'); at} :: es' ->
+    Lib.Option.map (fun c'' -> vs, {it = Frame (n, f, c''); at} :: es') (try_unsuspend c' m addr)
+  | {it = Suspend (m', addr', timeout); at} :: es' ->
+    if Memory.same m m' && addr = addr' then
+      Some (I32 0l :: vs, es')
+    else
+      None
+  | _ ->
+    None
+
+let rec wake (c : config) (m : memory_inst) (addr : Memory.address) (count : int32) : config * int32 =
+  if count = 0l then
+    c, 0l
+  else
+    match c with
+    | [] ->
+      c, 0l
+    | t :: ts ->
+      let t', count' = match (try_unsuspend t.code m addr) with | None -> t, 0l | Some c' -> {t with code = c'}, 1l in
+      let ts', count'' = wake ts m addr (Int32.sub count count') in
+      t' :: ts', Int32.add count' count''
 
 let rec step (c : config) (n : thread_id) : config =
   let ts1, t, ts2 = Lib.List.extract n c in
   if snd t.code = [] then
     step c n
   else
-    let t' = try step_thread t with Stack_overflow ->
+    let t', act = try step_thread t with Stack_overflow ->
       Exhaustion.error (List.hd (snd t.code)).at "call stack exhausted"
-    in ts1 @ [t'] @ ts2
+    in
+    match act with
+    | NotifyAction (m, addr, count) ->
+      let ts1', count1 = wake ts1 m addr count in
+      let ts2', count2 = wake ts2 m addr (Int32.sub count count1) in
+      ts1' @ [{t' with code = plug_value t'.code (I32 (Int32.add count1 count2))}] @ ts2'
+    | _ -> ts1 @ [t'] @ ts2
```

Comment:
If WaitAction is not distinguished here, why is it actually needed?

Reply:
Okay, we may want to think about getting the interpreter closer to the spec somehow wrt the actions it emits, and possibly how it uses them. But let's land this for now.